src/BackupRestore.ts

```diff
@@ -0,0 +1,56 @@
+import { keccak256 } from 'js-sha3';
+import nacl from 'ecma-nacl';
+import * as base64 from 'base64-arraybuffer';
+import { generateSecureRandom } from 'react-native-securerandom';
+
+import { hexToByteArray } from './Swarm';
+import { Version } from './Version';
+import * as utf8 from 'utf8-encoder';
+
+const header = `-----BEGIN FELFELE BACKUP-----`;
+const headerFields = `
+Version: Felfele v${Version}
+`;
+const separator = '\n\n';
+const footer = `-----END FELFELE BACKUP-----`;
+
+export const stripHeaderAndFooter = (backupText: string): string => {
+    if (!isValidBackup(backupText)) {
+        throw new Error('invalid backup header');
+    }
+    const backupDataParts = backupText.split(separator);
+    if (backupDataParts.length !== 3) {
+        throw new Error('invalid backup data parts');
+    }
+    return backupDataParts[1].replace('\n', '');
+};
+
+const wrapWithHeaderAndFooter = (data: string): string => {
+    return header + headerFields + separator + data + separator + footer;
+};
+
+export const isValidBackup = (backupText: string): boolean => {
+    if (!backupText.startsWith(header)) {
+        return false;
+    }
+    return true;
+};
+
+export const createBackupFromString = async (data: string, secretHex: string, generateRandom: (num: number) => Promise<Uint8Array> = generateSecureRandom): Promise<string> => {
+    const dataUint8Array = utf8.fromString(data);
+    const secureRandomUint8Array = await generateRandom(24);
+    const secretUint8Array = new Uint8Array(keccak256.array(hexToByteArray(secretHex)));
+    const encryptedData = nacl.secret_box.formatWN.pack(dataUint8Array, secureRandomUint8Array, secretUint8Array);
+    const encryptedDataBase64 = base64.encode(encryptedData);
+    const backupText = wrapWithHeaderAndFooter(encryptedDataBase64);
+    return backupText;
+};
+
+export const restoreBackupToString = async (backupText: string, secretHex: string): Promise<string> => {
+    const encryptedDataBase64 = stripHeaderAndFooter(backupText);
+    const secretUint8Array = new Uint8Array(keccak256.array(hexToByteArray(secretHex)));
+    const encryptedBytes = new Uint8Array(base64.decode(encryptedDataBase64));
```

Comment:
shouldn't base64.decode come after stripHeaderAndFooter?

Reply:
No, the header and footer is non-binary data that is useful to provide human readable information about the backup. For example we provide a version field that the app can use to decide if it is able to open the backup or not. With this it's possible to change for example the encoding to base85 or something else in a later version and it will be easy to handle that change from the code.

I used a similar model that PGP uses for encrypting emails, see https://pitt.edu/%7Epoole/PGP.htm#step8

Reply:
that's not what I meant, that is clear. I'm sure you tried it and works, but seemingly the order is not the reverse of the other function (the penultimate step is encode, so here the second step should be decode)

Reply:
The second step here is to process to passed in hex secret string for passing it to the encryption and decryption function, which is the same in both cases. It is not strictly part of the encryption process, however it's necessary to do it to get the code work. This could be potentially the zeroth step.

Reply:
oh I see it now, 👍